File: ekodb-client-kt/src/test/kotlin/io/ekodb/client/QueryBuilderTest.kt
package io.ekodb.client

import kotlinx.serialization.json.*
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Unit tests for ekoDB Kotlin client QueryBuilder
 */
class QueryBuilderTest {

    // ========================================================================
    // Basic Tests
    // ========================================================================

    @Test
    fun `creates empty query builder`() {
        val qb = QueryBuilder()
        assertNotNull(qb)
    }

    @Test
    fun `builds empty query`() {
        val query = QueryBuilder().build()
        assertNull(query.filter)
        assertNull(query.sort)
        assertNull(query.limit)
        assertNull(query.skip)
    }

    @Test
    fun `companion new creates instance`() {
        val qb = QueryBuilder.new()
        assertNotNull(qb)
    }

    // ========================================================================
    // Equality Operators Tests
    // ========================================================================

    @Test
    fun `builds eq filter`() {
        val query = QueryBuilder()
            .eq("status", "active")
            .build()

        assertNotNull(query.filter)
        val filter = query.filter as JsonObject
        assertEquals("Condition", filter["type"]?.jsonPrimitive?.content)
        val content = filter["content"]?.jsonObject
        assertEquals("status", content?.get("field")?.jsonPrimitive?.content)
        assertEquals("Eq", content?.get("operator")?.jsonPrimitive?.content)
        assertEquals("active", content?.get("value")?.jsonPrimitive?.content)
    }

    @Test
    fun `builds ne filter`() {
        val query = QueryBuilder()
            .ne("status", "deleted")
            .build()

        assertNotNull(query.filter)
        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Ne", content?.get("operator")?.jsonPrimitive?.content)
    }

    // ========================================================================
    // Comparison Operators Tests
    // ========================================================================

    @Test
    fun `builds gt filter`() {
        val query = QueryBuilder()
            .gt("age", 18)
            .build()

        assertNotNull(query.filter)
        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Gt", content?.get("operator")?.jsonPrimitive?.content)
        assertEquals(18, content?.get("value")?.jsonPrimitive?.int)
    }

    @Test
    fun `builds gte filter`() {
        val query = QueryBuilder()
            .gte("score", 80)
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Gte", content?.get("operator")?.jsonPrimitive?.content)
    }

    @Test
    fun `builds lt filter`() {
        val query = QueryBuilder()
            .lt("price", 100.5)
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Lt", content?.get("operator")?.jsonPrimitive?.content)
    }

    @Test
    fun `builds lte filter`() {
        val query = QueryBuilder()
            .lte("count", 1000)
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Lte", content?.get("operator")?.jsonPrimitive?.content)
    }

    // ========================================================================
    // Array Operators Tests
    // ========================================================================

    @Test
    fun `builds inArray filter`() {
        val query = QueryBuilder()
            .inArray("status", listOf("active", "pending"))
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("In", content?.get("operator")?.jsonPrimitive?.content)
        val values = content?.get("value")?.jsonArray
        assertEquals(2, values?.size)
    }

    @Test
    fun `builds nin filter`() {
        val query = QueryBuilder()
            .nin("role", listOf("blocked", "deleted"))
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("NotIn", content?.get("operator")?.jsonPrimitive?.content)
    }

    // ========================================================================
    // String Operators Tests
    // ========================================================================

    @Test
    fun `builds contains filter`() {
        val query = QueryBuilder()
            .contains("email", "@example.com")
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Contains", content?.get("operator")?.jsonPrimitive?.content)
        assertEquals("@example.com", content?.get("value")?.jsonPrimitive?.content)
    }

    @Test
    fun `builds regex filter`() {
        val query = QueryBuilder()
            .regex("phone", "^\\+1")
            .build()

        val filter = query.filter as JsonObject
        val content = filter["content"]?.jsonObject
        assertEquals("Regex", content?.get("operator")?.jsonPrimitive?.content)
    }

    // ========================================================================
    // Multiple Filters (Auto AND) Tests
    // ========================================================================

    @Test
    fun `combines multiple filters with AND logic`() {
        val query = QueryBuilder()
            .eq("status", "active")
            .gt("age", 18)
            .contains("email", "@company.com")
            .build()

        val filter = query.filter as JsonObject
        assertEquals("Logical", filter["type"]?.jsonPrimitive?.content)
        val content = filter["content"]?.jsonObject
        assertEquals("And", content?.get("operator")?.jsonPrimitive?.content)
        val expressions = content?.get("expressions")?.jsonArray
        assertEquals(3, expressions?.size)
    }

    // ========================================================================
    // Sorting Tests
    // ========================================================================

    @Test
    fun `builds ascending sort`() {
        val query = QueryBuilder()
            .sortAsc("name")
            .build()

        assertNotNull(query.sort)
        val sort = query.sort as JsonArray
        assertEquals(1, sort.size)
        val sortField = sort[0].jsonObject
        assertEquals("name", sortField["field"]?.jsonPrimitive?.content)
        assertEquals(true, sortField["ascending"]?.jsonPrimitive?.boolean)
    }

    @Test
    fun `builds descending sort`() {
        val query = QueryBuilder()
            .sortDesc("created_at")
            .build()

        val sort = query.sort as JsonArray
        val sortField = sort[0].jsonObject
        assertEquals(false, sortField["ascending"]?.jsonPrimitive?.boolean)
    }

    @Test
    fun `builds multiple sorts`() {
        val query = QueryBuilder()
            .sortDesc("created_at")
            .sortAsc("name")
            .build()

        val sort = query.sort as JsonArray
        assertEquals(2, sort.size)
        assertEquals("created_at", sort[0].jsonObject["field"]?.jsonPrimitive?.content)
        assertEquals(false, sort[0].jsonObject["ascending"]?.jsonPrimitive?.boolean)
        assertEquals("name", sort[1].jsonObject["field"]?.jsonPrimitive?.content)
        assertEquals(true, sort[1].jsonObject["ascending"]?.jsonPrimitive?.boolean)
    }

    @Test
    fun `sortAscending is alias for sortAsc`() {
        val query = QueryBuilder()
            .sortAscending("name")
            .build()

        val sort = query.sort as JsonArray
        assertEquals(true, sort[0].jsonObject["ascending"]?.jsonPrimitive?.boolean)
    }

    @Test
    fun `sortDescending is alias for sortDesc`() {
        val query = QueryBuilder()
            .sortDescending("name")
            .build()

        val sort = query.sort as JsonArray
        assertEquals(false, sort[0].jsonObject["ascending"]?.jsonPrimitive?.boolean)
    }

    // ========================================================================
    // Pagination Tests
    // ========================================================================

    @Test
    fun `builds limit`() {
        val query = QueryBuilder()
            .limit(10)
            .build()

        assertEquals(10, query.limit)
    }

    @Test
    fun `builds skip`() {
        val query = QueryBuilder()
            .skip(20)
            .build()

        assertEquals(20, query.skip)
    }

    @Test
    fun `builds limit and skip together`() {
        val query = QueryBuilder()
            .limit(10)
            .skip(20)
            .build()

        assertEquals(10, query.limit)
        assertEquals(20, query.skip)
    }

    // ========================================================================
    // Join Tests
    // ========================================================================

    @Test
    fun `builds join configuration with map`() {
        val joinConfig = mapOf(
            "collection" to "users",
            "local_field" to "user_id",
            "foreign_field" to "id",
            "as" to "user"
        )

        val query = QueryBuilder()
            .join(joinConfig)
            .build()

        assertNotNull(query.join)
    }

    @Test
    fun `builds join with joinWith helper`() {
        val query = QueryBuilder()
            .joinWith(
                collection = "users",
                localField = "user_id",
                foreignField = "id",
                alias = "user_data"
            )
            .build()

        assertNotNull(query.join)
        val join = query.join as JsonObject
        assertEquals("users", join["collection"]?.jsonPrimitive?.content)
        assertEquals("user_id", join["local_field"]?.jsonPrimitive?.content)
        assertEquals("id", join["foreign_field"]?.jsonPrimitive?.content)
        assertEquals("user_data", join["as"]?.jsonPrimitive?.content)
    }

    @Test
    fun `joinWith uses collection as default alias`() {
        val query = QueryBuilder()
            .joinWith(
                collection = "users",
                localField = "user_id",
                foreignField = "id"
            )
            .build()

        val join = query.join as JsonObject
        assertEquals("users", join["as"]?.jsonPrimitive?.content)
    }

    // ========================================================================
    // Bypass Flags Tests
    // ========================================================================

    @Test
    fun `builds bypass_cache true`() {
        val query = QueryBuilder()
            .bypassCache(true)
            .build()

        assertEquals(true, query.bypassCache)
    }

    @Test
    fun `builds bypass_cache default true`() {
        val query = QueryBuilder()
            .bypassCache()
            .build()

        assertEquals(true, query.bypassCache)
    }

    @Test
    fun `builds bypass_cache false`() {
        val query = QueryBuilder()
            .bypassCache(false)
            .build()

        assertEquals(false, query.bypassCache)
    }

    @Test
    fun `builds bypass_ripple`() {
        val query = QueryBuilder()
            .bypassRipple(true)
            .build()

        assertEquals(true, query.bypassRipple)
    }

    // ========================================================================
    // Chaining Tests
    // ========================================================================

    @Test
    fun `supports full method chaining`() {
        val query = QueryBuilder()
            .eq("status", "active")
            .gt("age", 18)
            .sortDesc("created_at")
            .sortAsc("name")
            .limit(10)
            .skip(20)
            .bypassCache(true)
            .build()

        // Check filter exists
        assertNotNull(query.filter)

        // Check sort exists
        val sort = query.sort as JsonArray
        assertEquals(2, sort.size)

        // Check pagination
        assertEquals(10, query.limit)
        assertEquals(20, query.skip)

        // Check bypass flag
        assertEquals(true, query.bypassCache)
    }

    @Test
    fun `returns this for method chaining`() {
        val qb = QueryBuilder()

        assertTrue(qb.eq("a", 1) === qb)
        assertTrue(qb.ne("b", 2) === qb)
        assertTrue(qb.gt("c", 3) === qb)
        assertTrue(qb.gte("d", 4) === qb)
        assertTrue(qb.lt("e", 5) === qb)
        assertTrue(qb.lte("f", 6) === qb)
        assertTrue(qb.inArray("g", listOf(7)) === qb)
        assertTrue(qb.nin("h", listOf(8)) === qb)
        assertTrue(qb.contains("i", "j") === qb)
        assertTrue(qb.regex("k", "l") === qb)
        assertTrue(qb.sortAsc("m") === qb)
        assertTrue(qb.sortDesc("n") === qb)
        assertTrue(qb.limit(1) === qb)
        assertTrue(qb.skip(0) === qb)
        assertTrue(qb.bypassCache() === qb)
        assertTrue(qb.bypassRipple() === qb)
    }

    // ========================================================================
    // SortOrder Enum Tests
    // ========================================================================

    @Test
    fun `SortOrder has correct values`() {
        assertEquals(SortOrder.ASC, SortOrder.valueOf("ASC"))
        assertEquals(SortOrder.DESC, SortOrder.valueOf("DESC"))
    }
}

File: ekodb-client-ts/src/query-builder.test.ts
/**
 * Unit tests for ekoDB TypeScript client QueryBuilder
 */

import { describe, it, expect } from "vitest";
import { QueryBuilder, SortOrder } from "./query-builder";

// ============================================================================
// Basic Tests
// ============================================================================

describe("QueryBuilder basics", () => {
  it("creates empty query builder", () => {
    const qb = new QueryBuilder();
    expect(qb).toBeInstanceOf(QueryBuilder);
  });

  it("builds empty query", () => {
    const query = new QueryBuilder().build();
    expect(query).toEqual({});
  });
});

// ============================================================================
// Equality Operators Tests
// ============================================================================

describe("QueryBuilder equality operators", () => {
  it("builds eq filter", () => {
    const query = new QueryBuilder().eq("status", "active").build();

    expect(query.filter).toEqual({
      type: "Condition",
      content: {
        field: "status",
        operator: "Eq",
        value: "active",
      },
    });
  });

  it("builds ne filter", () => {
    const query = new QueryBuilder().ne("status", "deleted").build();

    expect(query.filter.content.operator).toBe("Ne");
  });
});

// ============================================================================
// Comparison Operators Tests
// ============================================================================

describe("QueryBuilder comparison operators", () => {
  it("builds gt filter", () => {
    const query = new QueryBuilder().gt("age", 18).build();

    expect(query.filter.content.operator).toBe("Gt");
    expect(query.filter.content.value).toBe(18);
  });

  it("builds gte filter", () => {
    const query = new QueryBuilder().gte("score", 80).build();

    expect(query.filter.content.operator).toBe("Gte");
  });

  it("builds lt filter", () => {
    const query = new QueryBuilder().lt("price", 100.5).build();

    expect(query.filter.content.operator).toBe("Lt");
  });

  it("builds lte filter", () => {
    const query = new QueryBuilder().lte("count", 1000).build();

    expect(query.filter.content.operator).toBe("Lte");
  });
});

// ============================================================================
// Array Operators Tests
// ============================================================================

describe("QueryBuilder array operators", () => {
  it("builds in filter", () => {
    const query = new QueryBuilder()
      .in("status", ["active", "pending"])
      .build();

    expect(query.filter.content.operator).toBe("In");
    expect(query.filter.content.value).toEqual(["active", "pending"]);
  });

  it("builds nin filter", () => {
    const query = new QueryBuilder()
      .nin("role", ["blocked", "deleted"])
      .build();

    expect(query.filter.content.operator).toBe("NotIn");
  });
});

// ============================================================================
// String Operators Tests
// ============================================================================

describe("QueryBuilder string operators", () => {
  it("builds contains filter", () => {
    const query = new QueryBuilder().contains("email", "@example.com").build();

    expect(query.filter.content.operator).toBe("Contains");
    expect(query.filter.content.value).toBe("@example.com");
  });

  it("builds startsWith filter", () => {
    const query = new QueryBuilder().startsWith("name", "John").build();

    expect(query.filter.content.operator).toBe("StartsWith");
  });

  it("builds endsWith filter", () => {
    const query = new QueryBuilder().endsWith("filename", ".pdf").build();

    expect(query.filter.content.operator).toBe("EndsWith");
  });

  it("builds regex filter", () => {
    const query = new QueryBuilder().regex("phone", "^\\+1").build();

    expect(query.filter.content.operator).toBe("Regex");
  });
});

// ============================================================================
// Logical Operators Tests
// ============================================================================

describe("QueryBuilder logical operators", () => {
  it("builds and filter", () => {
    const conditions = [
      {
        type: "Condition",
        content: { field: "status", operator: "Eq", value: "active" },
      },
      {
        type: "Condition",
        content: { field: "age", operator: "Gt", value: 18 },
      },
    ];

    const query = new QueryBuilder().and(conditions).build();

    expect(query.filter.type).toBe("Logical");
    expect(query.filter.content.operator).toBe("And");
    expect(query.filter.content.expressions).toHaveLength(2);
  });

  it("builds or filter", () => {
    const conditions = [
      {
        type: "Condition",
        content: { field: "role", operator: "Eq", value: "admin" },
      },
      {
        type: "Condition",
        content: { field: "role", operator: "Eq", value: "super_admin" },
      },
    ];

    const query = new QueryBuilder().or(conditions).build();

    expect(query.filter.content.operator).toBe("Or");
  });

  it("builds not filter", () => {
    const condition = {
      type: "Condition",
      content: { field: "deleted", operator: "Eq", value: true },
    };

    const query = new QueryBuilder().not(condition).build();

    expect(query.filter.content.operator).toBe("Not");
  });
});

// ============================================================================
// Multiple Filters (Auto AND) Tests
// ============================================================================

describe("QueryBuilder multiple filters", () => {
  it("combines multiple filters with AND logic", () => {
    const query = new QueryBuilder()
      .eq("status", "active")
      .gt("age", 18)
      .contains("email", "@company.com")
      .build();

    expect(query.filter.type).toBe("Logical");
    expect(query.filter.content.operator).toBe("And");
    expect(query.filter.content.expressions).toHaveLength(3);
  });
});

// ============================================================================
// Sorting Tests
// ============================================================================

describe("QueryBuilder sorting", () => {
  it("builds ascending sort", () => {
    const query = new QueryBuilder().sortAsc("name").build();

    expect(query.sort).toHaveLength(1);
    expect(query.sort![0].field).toBe("name");
    expect(query.sort![0].ascending).toBe(true);
  });

  it("builds descending sort", () => {
    const query = new QueryBuilder().sortDesc("created_at").build();

    expect(query.sort![0].ascending).toBe(false);
  });

  it("builds multiple sorts", () => {
    const query = new QueryBuilder()
      .sortDesc("created_at")
      .sortAsc("name")
      .build();

    expect(query.sort).toHaveLength(2);
    expect(query.sort![0].field).toBe("created_at");
    expect(query.sort![0].ascending).toBe(false);
    expect(query.sort![1].field).toBe("name");
    expect(query.sort![1].ascending).toBe(true);
  });
});

// ============================================================================
// Pagination Tests
// ============================================================================

describe("QueryBuilder pagination", () => {
  it("builds limit", () => {
    const query = new QueryBuilder().limit(10).build();

    expect(query.limit).toBe(10);
  });

  it("builds skip", () => {
    const query = new QueryBuilder().skip(20).build();

    expect(query.skip).toBe(20);
  });

  it("builds page (convenience method)", () => {
    // Page 2 with 20 items per page = skip 40
    const query = new QueryBuilder().page(2, 20).build();

    expect(query.limit).toBe(20);
    expect(query.skip).toBe(40);
  });

  it("builds page 0", () => {
    const query = new QueryBuilder().page(0, 10).build();

    expect(query.skip).toBe(0);
    expect(query.limit).toBe(10);
  });
});

// ============================================================================
// Join Tests
// ============================================================================

describe("QueryBuilder join", () => {
  it("builds join configuration", () => {
    const joinConfig = {
      collections: ["users"],
      local_field: "user_id",
      foreign_field: "id",
      as_field: "user",
    };

    const query = new QueryBuilder().join(joinConfig).build();

    expect(query.join).toEqual(joinConfig);
  });
});

// ============================================================================
// Bypass Flags Tests
// ============================================================================

describe("QueryBuilder bypass flags", () => {
  it("builds bypass_cache true", () => {
    const query = new QueryBuilder().bypassCache(true).build();

    expect(query.bypass_cache).toBe(true);
  });

  it("builds bypass_cache false (not included)", () => {
    const query = new QueryBuilder().bypassCache(false).build();

    expect(query.bypass_cache).toBeUndefined();
  });

  it("builds bypass_cache default (true)", () => {
    const query = new QueryBuilder().bypassCache().build();

    expect(query.bypass_cache).toBe(true);
  });

  it("builds bypass_ripple", () => {
    const query = new QueryBuilder().bypassRipple(true).build();

    expect(query.bypass_ripple).toBe(true);
  });
});

// ============================================================================
// Chaining Tests
// ============================================================================

describe("QueryBuilder chaining", () => {
  it("supports full method chaining", () => {
    const query = new QueryBuilder()
      .eq("status", "active")
      .gt("age", 18)
      .sortDesc("created_at")
      .sortAsc("name")
      .limit(10)
      .skip(20)
      .bypassCache(true)
      .build();

    // Check filter exists
    expect(query.filter).toBeDefined();

    // Check sort exists
    expect(query.sort).toHaveLength(2);

    // Check pagination
    expect(query.limit).toBe(10);
    expect(query.skip).toBe(20);

    // Check bypass flag
    expect(query.bypass_cache).toBe(true);
  });

  it("returns this for method chaining", () => {
    const qb = new QueryBuilder();

    expect(qb.eq("a", 1)).toBe(qb);
    expect(qb.ne("b", 2)).toBe(qb);
    expect(qb.gt("c", 3)).toBe(qb);
    expect(qb.gte("d", 4)).toBe(qb);
    expect(qb.lt("e", 5)).toBe(qb);
    expect(qb.lte("f", 6)).toBe(qb);
    expect(qb.in("g", [7])).toBe(qb);
    expect(qb.nin("h", [8])).toBe(qb);
    expect(qb.contains("i", "j")).toBe(qb);
    expect(qb.startsWith("k", "l")).toBe(qb);
    expect(qb.endsWith("m", "n")).toBe(qb);
    expect(qb.regex("o", "p")).toBe(qb);
    expect(qb.sortAsc("q")).toBe(qb);
    expect(qb.sortDesc("r")).toBe(qb);
    expect(qb.limit(1)).toBe(qb);
    expect(qb.skip(0)).toBe(qb);
    expect(qb.bypassCache()).toBe(qb);
    expect(qb.bypassRipple()).toBe(qb);
  });
});

// ============================================================================
// Raw Filter Tests
// ============================================================================

describe("QueryBuilder rawFilter", () => {
  it("adds raw filter expression", () => {
    const rawFilter = {
      type: "Condition",
      content: {
        field: "custom",
        operator: "CustomOp",
        value: "custom_value",
      },
    };

    const query = new QueryBuilder().rawFilter(rawFilter).build();

    expect(query.filter).toEqual(rawFilter);
  });
});

// ============================================================================
// SortOrder Enum Tests
// ============================================================================

describe("SortOrder enum", () => {
  it("has correct values", () => {
    expect(SortOrder.Asc).toBe("asc");
    expect(SortOrder.Desc).toBe("desc");
  });
});
